virtio-devices/src/balloon.rs

```diff
@@ -165,6 +211,45 @@ impl BalloonEpollHandler {
         Self::advise_memory_range(memory, range_base, range_len, libc::MADV_DONTNEED)
     }
 
+    fn release_memory_range_4k(
+        pbp: &mut Option<PartiallyBalloonedPage>,
+        memory: &GuestMemoryMmap,
+        pfn: u32,
+    ) -> result::Result<(), Error> {
+        let range_base = GuestAddress((pfn as u64) << VIRTIO_BALLOON_PFN_SHIFT);
+        let range_len = 1 << VIRTIO_BALLOON_PFN_SHIFT;
+
+        let page_size: u64 = get_page_size();
+        if page_size == 1 << VIRTIO_BALLOON_PFN_SHIFT {
+            return Self::release_memory_range(memory, range_base, range_len);
+        }
+
+        if pbp.is_none() {
+            *pbp = Some(PartiallyBalloonedPage::new());
+        }
+
+        if !pbp.as_ref().unwrap().pfn_match(range_base.0) {
+            // We are trying to free memory region in a different pfn with current pbp. Flush pbp.
+            pbp.as_mut().unwrap().reset();
+            pbp.as_mut().unwrap().addr = align_page_size_down(range_base.0);
+        }
+
+        pbp.as_mut().unwrap().set_bit(range_base.0);
+        if !pbp.as_ref().unwrap().bitmap_full() {
+            return Ok(());
+        }
+
+        Self::release_memory_range(
+            memory,
+            vm_memory::GuestAddress(pbp.as_ref().unwrap().addr),
+            page_size as usize,
+        )?;
+
+        pbp.as_mut().unwrap().reset();
+
+        Ok(())
```

Comment:
Nit: the code would be simpler if you do:

```
       if pbp.as_ref().unwrap().bitmap_full() {
            Self::release_memory_rang(
            ...
        }
        
       Ok(())

```

Reply:
OK